### PR TITLE
Painless: Types Section Clean Up

### PR DESCRIPTION
**Pull request by jdconrad**
Cleaned up the types and casting sections of the painless spec including improved definitions, improved examples, and improved general structure.

### CODE REVIEW

File: docs/painless/painless-casting.asciidoc [hunk 1]
```diff
@@ -1,172 +1,472 @@
...
-numeric types. Read the table by row. To find out if you need to explicitly
-cast from type A to type B, find the row for type A and scan across to the
-column for type B.
+A <<primitive-types, numeric type>> cast is the conversion of the value of one
+<<primitive-types, numeric type>> to the equivalent value of an inferred or
+specified <<primitive-types, numeric type>>. Casts between values of
+<<primitive-types, numeric types>> will result in data loss when the value of
+the type converted from is larger than the equivalent value of the type
+converted to can accommodate. Casts between values of integral types and values
+of floating point types will result in precision loss when the value of the
```
Comment (debadair) [inline docs/painless/painless-casting.asciidoc]: Maybe, "larger than the target type can accommodate"?

I'd drop the "wills" -- they do result in data/precision loss in those cases.

Maybe just "Casts between integer and floating point values can result in precision loss."?

File: docs/painless/painless-operators.asciidoc [hunk 2]
```diff
@@ -248,9 +249,49 @@ initialize each dimension with are specified as a comma-separated list enclosed
 in braces. For example, `new int[] {1, 2, 3}` creates a one dimensional `int`
 array with a size of 3 and the values 1, 2, and 3.
 
-For more information about allocating and initializing arrays, see <<array-type,
-Array Type>>.
+To allocate an array, you use the `new` keyword followed by the type and a
+set of brackets for each dimension. You can explicitly define the size of each dimension by specifying an expression within the brackets, or initialize each
```
Comment (debadair) [inline docs/painless/painless-operators.asciidoc]: I'd drop the "you" and just use the imperative: "To allocate an array, use..."

File: docs/painless/painless-operators.asciidoc [hunk 3]
```diff
@@ -248,9 +249,49 @@ initialize each dimension with are specified as a comma-separated list enclosed
...
+dimension with the desired number of values. The allocated size of each
+dimension is its permanent size.
 
+To initialize an array, specify the values you want to initialize
+each dimension with as a comma-separated list of expressions enclosed in braces.
+For example, `new int[] {1, 2, 3}` creates a one-dimensional `int` array with a
```
Comment (debadair) [inline docs/painless/painless-operators.asciidoc]: I'd cut "specify the values you want to initialize each dimension with as".

**Review by debadair — COMMENTED** [review]
I added a few editorial comments. I don't see any structural issues.

File: docs/painless/painless-casting.asciidoc [hunk 4]
```diff
@@ -1,172 +1,472 @@
 [[painless-casting]]
 === Casting
 
-Casting is the conversion of one type to another. Implicit casts are casts that
-occur automatically, such as during an assignment operation. Explicit casts are
-casts where you use the casting operator to explicitly convert one type to
-another. This is necessary during operations where the cast cannot be inferred.
+A cast is the conversion of the value of one type to the equivalent value of an
+inferred or specified type. Implicit casts are an inferred operation that
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply docs/painless/painless-casting.asciidoc]: Fixed.

File: docs/painless/painless-casting.asciidoc (hunk 1 above)
Comment (jdconrad) [reply docs/painless/painless-casting.asciidoc]: Fixed.

**Comment by jdconrad** [comment]
@debadair Thanks for the first round of review!  Took another pass and did my best to clean up based on your comments.  Ready for the next round.

**Comment by debadair** [comment]
LGTM!

**Comment by jdconrad** [comment]
@debadair Thank you for the review!  I realize this is quite time consuming to look through.

**Pull request merged**